src/Microsoft.AspNetCore.Hosting/Internal/RequestServicesContainerMiddleware.cs

```diff
@@ -25,37 +26,37 @@ public RequestServicesContainerMiddleware(RequestDelegate next, IServiceScopeFac
                 throw new ArgumentNullException(nameof(scopeFactory));
             }
 
-            _scopeFactory = scopeFactory;
             _next = next;
+            _scopeFactory = scopeFactory;
         }
 
         public async Task Invoke(HttpContext httpContext)
         {
-            if (httpContext == null)
-            {
-                throw new ArgumentNullException(nameof(httpContext));
-            }
+            Debug.Assert(httpContext != null);
 
-            var existingFeature = httpContext.Features.Get<IServiceProvidersFeature>();
+            // local cache for virtual disptach result
+            var features = httpContext.Features;
+            var existingFeature = features.Get<IServiceProvidersFeature>();
 
-            // All done if request services is set
+            // All done if RequestServices is set
             if (existingFeature?.RequestServices != null)
             {
                 await _next.Invoke(httpContext);
                 return;
             }
 
-            using (var feature = new RequestServicesFeature(_scopeFactory))
+            var replacementFeature = new RequestServicesFeature(_scopeFactory);
+
+            try
+            {
+                features.Set<IServiceProvidersFeature>(replacementFeature);
+                await _next.Invoke(httpContext);
+            }
+            finally
             {
-                try
-                {
-                    httpContext.Features.Set<IServiceProvidersFeature>(feature);
-                    await _next.Invoke(httpContext);
-                }
-                finally
-                {
-                    httpContext.Features.Set(existingFeature);
-                }
+                replacementFeature.Dispose();
```

Comment:
I guess `IFeatureCollection` *is* an abstraction and in theory it could throw. If you were relying on the order of dispose vs set it would probably be broken in other subtle ways... 😄